src/shared/api/memoize.ts

```diff
@@ -0,0 +1,155 @@
+ */
+class Cache<K,V> {
+	private map = new Map<any, V>();
+	private weakMap = new WeakMap<any, V>();
+
+	has(key:K):boolean {
+		return isPrimitive(key) ? this.map.has(key) : this.weakMap.has(key);
+	}
+	get(key:K):V|undefined {
+		return isPrimitive(key) ? this.map.get(key) : this.weakMap.get(key);
+	}
+	set(key:K, value:V):this {
+		if (isPrimitive(key))
+			this.map.set(key, value);
+		else
+			this.weakMap.set(key, value);
+		return this;
+	}
+}
+
+/**
+ * Provides a multi-dimensional Map-like interface
+ */
+class HyperMap<T> {
+	has(args:any[]):boolean {
+		let cache:Cache<any, any>|undefined = this.getCache(args.length);
+		if (args.length > 1)
+			cache = this.traverse(cache, args.slice(0, args.length - 1)) as any;
+		return cache !== undefined && cache.has(args[args.length - 1]);
+	}
+
+	get(args:any[]):T | undefined {
+		return this.traverse(this.getCache(args.length), args);
+	}
+
+	set(args:any[], value:T):void {
+		this.traverse(this.getCache(args.length), args, value);
+	}
+
+	// gets the Cache designated for a specific key length
+	private getCache(numArgs:number)
+	{
+		let cache = this.map_numArgs_cache.get(numArgs);
+		if (!cache)
+			this.map_numArgs_cache.set(numArgs, cache = new Cache());
+		return cache;
+	}
+
+	// used to avoiding the varying-key-length limitation of the traverse() function below
+	private map_numArgs_cache = new Map<number, Cache<any, any>>();
+
+	// dual-purpose setter/getter
+	// note: does not work if subsequent calls vary the length of the keys array for the same cache param
+	private traverse(cache:Cache<any, any>, keys:any[], value?:T):T|undefined {
+		if (keys.length == 0)
+			return undefined;
+		if (keys.length == 1) {
+			if (value === undefined)
+				return cache.get(keys[0]);
+			return void cache.set(keys[0], value);
+		}
+
+		let nextCache = cache.get(keys[0]);
+		if (nextCache === undefined)
+			cache.set(keys[0], nextCache = new Cache());
+		return this.traverse(nextCache, keys.slice(1), value);
+	}
+}
+
+class Memoizer<Result, Transform extends (...args:any[])=>Result> {
+
+	private transform:Transform;
+	private cache = new HyperMap<Result>();
+
+	constructor(transform:Transform, getAdditionalArgs?:()=>any[]) {
+		this.transform = transform;
+		let cache = this.cache;
+		this.get = function(...args:any[]) {
+			let thisArgs = [this];
+			if (getAdditionalArgs)
+				thisArgs = thisArgs.concat(getAdditionalArgs.call(this), args);
+			else
+				thisArgs = thisArgs.concat(args);
+			let result = cache.get(thisArgs);
+			if (result === undefined && !cache.has(thisArgs))
+			{
+				result = transform.apply(this, args);
+				if (result !== undefined)
+					cache.set(thisArgs, result);
+			}
+			return result;
+		} as Transform;
+	}
+
+	get:Transform;
+}
+
+/**
+ * Creates a memoized version of a function.
+ */
+export default function memoize<F extends (...args:any[])=>any>(fn:F, getAdditionalMemoizeArgs:()=>any[]):F;

```

Comment:
Could we use reselect instead of this?  Reselect is documented, tested and widely used.   https://github.com/reactjs/reselect.  I mean, if there's a good reason to write our own, then by all means.